File: Python/LeetCode/LeetCode1391.py
# https://leetcode.com/problems/check-if-there-is-a-valid-path-in-a-grid

from collections import deque
class Solution:
    def hasValidPath(self, grid: List[List[int]]) -> bool:
        m, n = len(grid), len(grid[0])
        
        def boundaryCheck(row: int, col: int) -> bool:
            return 0 <= row < m and 0 <= col < n
        
        graph = {}
        for row in range(m):
            for col in range(n):
                graph[(row, col)] = []
                if grid[row][col] == 1:
                    candidates = [(row, col - 1), (row, col + 1)]
                elif grid[row][col] == 2:
                    candidates = [(row - 1, col), (row + 1, col)]
                elif grid[row][col] == 3:
                    candidates = [(row, col - 1), (row + 1, col)]
                elif grid[row][col] == 4:
                    candidates = [(row, col + 1), (row + 1, col)]
                elif grid[row][col] == 5:
                    candidates = [(row, col - 1), (row - 1, col)]
                else:
                    candidates = [(row, col + 1), (row - 1, col)]
                for nr, nc in candidates:
                    if boundaryCheck(nr, nc):
                        graph[(row, col)].append((nr, nc))
        queue = deque([(0, 0)])
        visited = [[False] * n for _ in range(m)]
        visited[0][0] = True
        while queue:
            row, col = queue.popleft()
            if row == m - 1 and col == n - 1:
                return True
            for nr, nc in graph[(row, col)]:
                if visited[nr][nc]:
                    continue
                if (row, col) not in graph[(nr, nc)]:
                    continue
                visited[nr][nc] = True
                queue.append((nr, nc))
        return False

File: swift/leetcode/leetcode1391.swift
class Solution {
    func hasValidPath(_ grid: [[Int]]) -> Bool {
        struct Queue {
            private var container = [(Int, Int)]()
            private var ptr = 0
            var isEmpty: Bool {
                ptr >= container.count
            }
            mutating func push(_ val: (Int, Int)) {
                container.append(val)
            }
            mutating func pop() -> (Int, Int) {
                let frontValue = container[ptr]
                ptr += 1
                if ptr > 100000 {
                    container = Array(container[ptr...])
                    ptr = 0
                }
                return frontValue
            }
        }
        
        func bfs() -> Bool {
            var queue = Queue()
            queue.push((0, 0))
            var isVisited = [[Bool]](repeating: [Bool](repeating: false, count: n), count: m)
            isVisited[0][0] = true
            let directions = [
                1: [(0, -1), (0, 1)],
                2: [(-1, 0), (1, 0)],
                3: [(0, -1), (1, 0)],
                4: [(0, 1), (1, 0)],
                5: [(0, -1), (-1, 0)],
                6: [(0, 1), (-1, 0)]
            ]
            while !queue.isEmpty {
                let (row, col) = queue.pop()
                for direction in directions[grid[row][col]]! {
                    let (newRow, newCol) = (row + direction.0, col + direction.1)
                    guard 0..<m ~= newRow && 0..<n ~= newCol else { continue }
                    guard !isVisited[newRow][newCol] else { continue }
                    let reverse = (-direction.0, -direction.1)
                    guard directions[grid[newRow][newCol]]!.contains(where: {
                        $0.0 == reverse.0 && $0.1 == reverse.1
                    }) else {
                        continue
                    }
                    if newRow == m - 1 && newCol == n - 1 { return true }
                    isVisited[newRow][newCol] = true
                    queue.push((newRow, newCol))
                }
            }
            return false
        }
        let m = grid.count
        let n = grid.first?.count ?? 0
        guard m != 1 || n != 1 else { return true }
        return bfs()
    }
}
